client/starwhale/base/uricomponents/resource.py

```diff
@@ -0,0 +1,143 @@
+            accept patterns:
+                * resource name without type, e.g. mnist
+                * version e.g. gntdqm3fgeydqnrtmftdgyjzpj4hamy
+                * full uri without version e.g. local/project/self/dataset/mnist
+                * full uri e.g. local/project/self/dataset/mnist/version/latest
+                * resource name and version e.g. mnist/version/latest
+            typically unacceptable cases:
+                * resource type and name under project e.g. self/dataset/mnist
+                * resource type, name and version under project e.g. self/dataset/mnist/version/latest
+        :param typ: resource type: (dataset, model, runtime etc)
+        :param project: project which the resource belongs to (optional)
+        :return: Resource instance
+        """
+
+        if project:
+            self.project = project
+        else:
+            try:
+                self.project = Project.parse(uri)
+                uri = self.project.path
+            except ParseErrException:
+                self.project = Project()
+
+        parts = len(uri.split("/"))
+        # 3 == len('mnist/version/latest'.split(/))
+        # means that there is no type info in the uri
+        if typ and parts <= 3:
+            self._parse_with_type(typ, uri)
+        else:
+            # uri is long enough to parse type
+            self._parse_without_type(uri)
+        if (typ is not None) and self.typ != typ:
+            raise Exception(f"type mismatch: {self.typ} vs {typ}")
+
+    def _parse_with_type(self, typ: ResourceType, uri: str) -> None:
+        """
+        :param typ:
+        :param uri:
+            i.e.
+            * mnist
+            * version/latest
+            * mnist/version/latest
+        """
+        self.typ = typ
+        parts = uri.split("/")
+        if len(parts) > 3:
+            raise Exception(f"invalid uri {uri} when parse with type {typ}")
+
+        if len(parts) == 1:
+            self.name = parts[0]
+        elif len(parts) == 2:
+            if parts[0] != "version":
+                raise Exception(f"invalid uri {uri}")
+            self.version = parts[-1]
+        else:
+            if parts[1] != "version":
+                raise Exception(f"invalid uri {uri}")
+            self.name = parts[0]
+            self.version = parts[-1]
+
+    def _parse_without_type(self, uri: str) -> None:
+        p = uri.split("/", 1)
+        if len(p) == 1:
+            # version only (we do not support name without type parsing)
+            # there is no scenario for 'name without type' for now
+            return self._parse_by_version(uri)
+        typ = ResourceType[p[0]]
+        return self._parse_with_type(typ, p[1])
+
+    def _parse_by_version(self, ver: str) -> None:
+        if self.instance.is_local:
+            root = Path(load_swcli_config()["storage"]["root"]) / self.project.name
+            # storage-root/project/type/name/prefix/full-version
```

Comment:
should the function take account of the `recover path`?

a removed runtime example
```shell
/home/tianwei/.starwhale/self/runtime/.recover/pytorch-starwhale-main/ge/geydgnrqgbswcy3ehezggnjqorqwk5y.swrt
```

Reply:
When users use the version only, we assume they use the full version or the long enough version.
e.g. abcde, this will try to match `/home/tianwei/.starwhale/self/runtime/.recover/pytorch-starwhale-main/abcde*`
and it will match nothing.